### ISSUE TITLE
Cross platform fixes for WinUAE

### ISSUE THREAD

**Issue by emoon** [tonioni/WinUAE#285]
Hi,

Lots of the code in WinUAE is cross-platform today, but there are some cases here and there where the general code doesn't compile. This is excluding the Windows specific code of course.

I wonder if you would be open for PRs fixing these? These fixes will be made as least intrusive as possible and in many case they are straightforward, but some of them may require a bit more talking.

The plan would be to start with the easiest fixes first and then potentially have some more discussion about the more involved ones if they make sense or not.

How does that sound?

**Comment by midwan** [comment]
I can probably help with this as well, since Amiberry has those pieces of code altered already, in order to compile for Linux and macOS natively.

**Comment by tonioni** [comment]
Yes but it depends. Anything under pcem-directory won't be accepted. And no "gcc-isms" either.

**Comment by johnnovak** [comment]
Would be nice to make the UI and core emulator layers modular. Then ordinary mortal as myself could take a stab at writing some other alternative frontends for Linux/macOS without going near the core stuff 😅

**Comment by luvwagn** [comment]
Did the Linux version fork?  UAE used to be quite portable, and the Win parts were somewhat layered.
________________________________

**Comment by rofl0r** [comment]
there is no "the" linux version, at least not anymore. UAE itself is dead since decades. all that's left is a couple forks that try (or tried) to keep pace with WinUAE. PUAE tried to stay in C while still backporting WinUAE changes, until so much C++ crept in that it became inpractical so it's stuck somewhere around 2.8 level. FS-UAE went the path of shiny 3d gui, and afaik hasnt been updated to latest WinUAE for quite some time too. the best option currently is probably the amiberry version, even though it also lacks a lot of WinUAE's features.

**Comment by emoon** [comment]
> Would be nice to make the UI and core emulator layers modular too. Then ordinary mortal such as myself could take a stab at writing some other alternative frontends for Linux/macOS without going near the core stuff 😅
But then you'd need some abstract interface for all the audio, input handling, and shader stuff for that.

It is actually fairly modular, but there in no such thing as an interface. Most platform specific code for WinUAE is in `od-win32` and if you remove that then you are left with the functions you need to implement. Would a proper interface for this be easier and more clear to implement? For sure, but I'm not sure it's worth it.

> Did the Linux version fork?  UAE used to be quite portable, and the Win parts were somewhat layered.

This is mostly the case, but there are some instances when the general code doesn't compile on non-msvc anymore and these are the parts I plan to do PRs for at some point.

**Comment by johnnovak** [comment]
> Most platform specific code for WinUAE is in `od-win32` and if you remove that then you are left with the functions you need to implement.

Yeah, that sounds like all we need then. I should have said "proper separation" instead; I don't actually care if there's technically a literal interface there or not ;)

**Comment by Lucretia** [comment]
I would interested in an up to date cross platform UI for this. Is all the windows ui stuff inside od-win32?

The amiberry UI is like something from early 90's unix when people were trying to copy motif and failing, also, it's really unstable. And as others have said FS is dead, that that I liked it much anyway.

**Comment by midwan** [comment]
> I would interested in an up to date cross platform UI for this. Is all the windows ui stuff inside od-win32?
> 
> The amiberry UI is like something from early 90's unix when people were trying to copy motif and failing, also, it's really unstable. And as others have said FS is dead, that that I liked it much anyway.

The Amiberry UI is eventually going to be rewritten in something more modern, but I haven't had time to go into that yet. Feel free to send PRs if you can. The layout is designed to replicate WinUAE as much as possible, to keep things familiar.
Regarding instability though, please open issues if you find such things, to help improve/fix things.

**Comment by Lucretia** [comment]
I have no bugs to report, I tried it last year and uninstalled it.

What does "od" mean in "od-win32"?

**Comment by Lucretia** [comment]
> The Amiberry UI is eventually going to be rewritten in something more modern

I didn't realise you use SDL. I would recommend something like dear imgui or nuklear.

**Comment by rofl0r** [comment]
> The amiberry UI is like something from early 90's unix

sounds more like a feature than a bug to me. modern is often inferior to classic. though in amiberry's gui there are still a few rough edges, the most important one probably that it's optimized for a hi-res screen and is almost unreadable at lower resolutions like 1024x768.

> What does "od" mean in "od-win32"?

my guess is "OS dependency". though not even toni might know for sure since IIRC i've seen the od- prefix even in 20 year old original UAE code.

**Comment by Lucretia** [comment]
> > The amiberry UI is like something from early 90's unix
> 
> sounds more like a feature than a bug to me. modern is often inferior to classic. though in amiberry's gui there are still a few rough edges, the most important one probably that it's optimized for a hi-res screen and is almost unreadable at lower resolutions like 1024x768.

Tell me about it, I can't read any of it at the size it opens.
 
> > What does "od" mean in "od-win32"?
> 
> my guess is "OS dependency". though not even toni might know for sure since IIRC i've seen the od- prefix even in 20 year old original UAE code.

Ah, didn't think it went back that far, thought it was winuae specific.

**Comment by tonioni** [comment]
Fixes been added from Amiberry and https://github.com/reinauer/WinUAE (and probably more comes later)
